### ISSUE TITLE
MultiTaskQuestionGeneratorForDecoder中的离散采样操作导致梯度无法回传

### ISSUE THREAD

**Issue by Curry-AI** [shiquanyang/NS-Dial#1]
作者你好，非常感谢你们的工作，论文idea和效果都令人惊艳，但我有一些疑问

你们的方法中涉及到三处离散采样，并使用了Gumbel softmax进行近似采样并使得梯度可以回传，但在Debug代码时发现，梯度并没有按照预期一样回传，部分参数的梯度始终为None，未进行更新，我不知道是不是我的理解有误，可以请作者解答一下吗？

如MultiTaskQuestionGeneratorForDecoder类的compute_structure_type方法

`   
 def compute_structure_type(self, hidden):
        sp_h = self.W1(hidden)  # [batch_size, 2]
        # sample an action
        sp_h_action = torch.nn.functional.gumbel_softmax(logits=sp_h, tau=1.0, hard=True) # hard gumbel softmax实现离散采样和梯度回传
        loss = None
        return sp_h, sp_h_action, loss
`
但是在loss.backward()后，self.W1.weight.grad依然是None，并没有对这部分的参数进行更新
